java/src/jmri/jmrix/openlcb/swing/memtool/MemoryToolPane.java

```diff
@@ -0,0 +1,427 @@
+                                        cbr);
+                } else {
+                    setRunning(false);
+                    cancelled = false;
+                    log.debug("Get operation cancelled");
+                    statusField.setText("Cancelled");
+                }
+            }
+        };
+
+    OutputStream outputStream;
+    long endingAddress = 0x1000; // token 1MB max if decide not to enquire about it & other methods fail
+
+    /**
+     * Starts reading from node and writing to file process
+     * @param e not used
+     */
+    void pushedGetButton(ActionEvent e) {
+        setRunning(true);
+        farID = nodeSelector.getSelectedItem();
+        try {
+            space = Integer.parseInt(spaceField.getText().trim());
+        } catch (NumberFormatException ex) {
+            log.error("error parsing the space field value \"{}\"", spaceField.getText());
+            statusField.setText("Error parsing the space value");
+            setRunning(false);
+            return;
+        }
+        log.debug("Start put");
+        if (fileChooser == null) {
+            fileChooser = new JFileChooser();
+        }
+        fileChooser.setDialogTitle("Read into binary file");
+        fileChooser.rescanCurrentDirectory();
+        fileChooser.setSelectedFile(new File("memory.bin"));
+
+        int retVal = fileChooser.showSaveDialog(this);
+        if (retVal != JFileChooser.APPROVE_OPTION) {
+            setRunning(false);
+            return;
+        }
+
+        // open file
+        File file = fileChooser.getSelectedFile();
+        log.debug("access {}", file);
+        try {
+            outputStream = new FileOutputStream(file);
+        } catch (IOException ex) {
+            log.error("Error opening file", ex);
+            statusField.setText("Error opening file");
+            setRunning(false);
+            return;
+        }
+
+        if (trustStatusReply.isSelected()) {
+            // request address space info; reply will start read operations.
+            // Memo has to be created here to carry appropriate farID
+            MemoryConfigurationService.McsAddrSpaceMemo cbq =
+                new MemoryConfigurationService.McsAddrSpaceMemo(farID, space) {
+                    @Override
+                    public void handleWriteReply(int errorCode) {
+                        log.error("Get failed with code {}", String.format("%04X", errorCode));
+                        statusField.setText("Get failed with code"+String.format("%04X", errorCode));
+                        setRunning(false);
+                    }
+
+                    @Override
+                    public void handleAddrSpaceData(NodeID dest, int space, long hiAddress, long lowAddress, int flags, String desc) {
+                        // check contents
+                        log.debug("received length of {}", hiAddress);
+                        endingAddress = hiAddress;
```

Comment:
Right now, I'm assuming that the starting address is zero, instead of using the low address returned from the Address Space inquiry. I can look into changing that, but I don't have any devices that provide a non-zero start address for testing AFAIK.

Reply:
Now handles non-zero low address for the starting location.